src/metabase/task/sweep_query_analysis.clj

```diff
@@ -0,0 +1,90 @@
+(ns metabase.task.sweep-query-analysis
+  (:require
+   [clojurewerkz.quartzite.jobs :as jobs]
+   [clojurewerkz.quartzite.schedule.cron :as cron]
+   [clojurewerkz.quartzite.triggers :as triggers]
+   [metabase.query-analysis :as query-analysis]
+   [metabase.task :as task]
+   [toucan2.core :as t2])
+  (:import
+   (org.quartz DisallowConcurrentExecution)))
+
+(set! *warn-on-reflection* true)
+
+(defn- analyze-cards-without-query-fields!
+  ([]
+   (analyze-cards-without-query-fields! query-analysis/analyze-sync!))
+  ([analyze-fn]
+   (let [cards (t2/reducible-select [:model/Card :id]
+                                    {:left-join [[:query_field :f] [:= :f.card_id :report_card.id]]
+                                     :where     [:and
+                                                 [:not :report_card.archived]
+                                                 [:= :f.id nil]]})]
+     (run! analyze-fn cards))))
+
+(defn- analyze-stale-cards!
+  ([]
+   (analyze-cards-without-query-fields! query-analysis/analyze-sync!))
+  ([analyze-fn]
+   (let [cards (t2/reducible-select [:model/Card :id])]
+     (run! analyze-fn cards))))
+
+(defn- delete-orphan-analysis! []
+  (transduce
+   (comp (map :id)
+         (partition-all 3))
+   (fn
+     ([final-count] final-count)
+     ([running-count ids]
+      (t2/delete! :model/QueryField :id [:in ids])
+      (+ running-count (count ids))))
+   0
+   (t2/reducible-select [:model/QueryField :id]
+                        {:join  [[:report_card :c] [:= :c.id :query_field.card_id]]
+                         :where [:or
+                                 [:= :c.id nil]
+                                 :c.archived]})))
+
+(def ^:private has-run?
+  "Has the sweeper been run before?"
+  (atom false))
+
+(defn- sweep-query-analysis-loop!
+  ([]
+   (sweep-query-analysis-loop! (not @has-run?))
+   (reset! has-run? true))
+  ([first-time?]
+   (sweep-query-analysis-loop! first-time? query-analysis/analyze-sync!))
+  ([first-time? analyze-fn]
+   ;; prioritize cards that are missing analysis
+   (analyze-cards-without-query-fields! analyze-fn)
+
+   ;; we run through all the existing analysis on our first run, as it may be stale due to an old macaw version, etc.
+   (when first-time?
```

Comment:
I'm on the fence about this. The problem is that if we drop an event, we're left with stale analysis until the instance restarts.

Reply:
without the timestamp column is there a reasonable way around it?

Reply:
Well, the alternative is to keep looping through everything, each time. That makes me uneasy.